staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go

```diff
@@ -96,9 +96,8 @@ var (
 			Help: "Response latency distribution in seconds for each verb, dry run value, group, version, resource, subresource, scope and component.",
 			// This metric is used for verifying api call latencies SLO,
 			// as well as tracking regressions in this aspects.
-			// Thus we customize buckets significantly, to empower both usecases.
-			Buckets: []float64{0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0,
-				1.25, 1.5, 1.75, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5, 6, 7, 8, 9, 10, 15, 20, 25, 30, 40, 50, 60},
+			// Thus we like to have good resolution between 0.1s and 1s.
+			Buckets:        []float64{0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.5, 2.0, 4.0, 10, 20, 40, 60},
```

Comment:
I still strongly believe that this isn't enough to debug and understand what is happening:
- above 1s, we at least need 5s and 30s buckets, because those are the numbers where SLO thresholds are for some type of requests: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md
so I would suggest:
1, 1.5, 2, 2.5, 5, 7.5, 10, 15, 20, 30, 45, 60

- below 1s, it's still too low granularity and added all steps by 0.1

Reply:
Thanks for the reference for the SLOs; I can add that to the comment.

Do you have one for "below 1s, it's still too low granularity"?  You seem to expect a very low standard distribution on call times; is that based on data I can cite?

I have found that Prometheus metrics are always too coarse to really understand a performance problem. I turn to distributed tracing when I need to dig deeper.

Reply:
@kubernetes/sig-scalability-misc @kubernetes/sig-scalability-pr-reviews

Reply:
@mm4tt - FYI

Reply:
I agree the existing number of buckets is probably excessive.

I think I'd like a 30s bucket (some calls time out at 30s and others at 32s). For > 10s range, I suggest: 10, 30, 60. If that's not enough detail, I'd go to: 10, 15, 21, 30, 42, 60 (roughly a factor of sqrt 2 apart, for even sizing).

for 1-10 second range, why not 1, 2, 5, 10?

likewise for sub-second range, .05, .1, .2, .5, 1?

Reply:
@logicalhan may have thoughts on this, but he's OOO at the moment.

Reply:
@lavalamp - what you suggest is waaaay to low granularity to understand and debug performance and regressions.

Reply:
Does it keep a count per bucket, or a count and a sum (i.e., an average)?

Reply:
So the thing that jumps out to me about all of these buckets -- the old ones and every proposed set on this thread -- is that they are obviously designed for human readability of the bucket boundaries themselves, which in practice hardly anybody will ever care about outside code reviews of this line. What you actually want is a distribution of buckets that makes simple interpolation an easy and reasonably sound way to get a sense of the actual distribution.

Traditionally in systems I've operated, we use exponentially-sized buckets, with ratios sized to give the tradeoff between fine-grained reporting and memory usage that makes sense for the system and metric in question; common values are 4 (for when only a very rough sense of order of magnitude is required), square root of 2, or fifth root of ten (the latter having the nice property that every fifth bucket is pretty close to a power of ten).

Perhaps more important than picking a single set of values here, though, is making it possible for individual system operators to make their own tradeoff here. I'm new to this codebase, so forgive the newbie question, but is there a reason this can't be configurable?

Reply:
If it only keeps a count, then I agree that more buckets are in order. I also agree w/ everything @jktomer said.

Not every cluster needs high resolution, only some clusters need to debug scalability regressions.

I also think that, even if the purpose is scalability confirmation/debugging, buckets should be exponentially distributed, not linearly-- a change from 1 to .9 is just 10%, a change from .2 to .1 is 50%.